### ISSUE TITLE
Generation Not Stopping (EOS Token & SQL Extraction Improvements)

### ISSUE THREAD

**Issue by chrisjcc** [chrisjcc/text-to-sql-finetuning#9]
**Context:**  
This issue continues the discussion from PR [#8: Generation Not Stopping (EOS Token Problem)](https://github.com/chrisjcc/text-to-sql-finetuning/pull/8).  
Assigning this task to **@claude** for implementation.

---

## 🎯 Objective

Improve SQL generation and post-processing logic in `scripts/evaluate.py`, ensuring outputs stop cleanly and match ground-truth SQL more reliably.  
This includes clarifying the extraction process and making the generation step more robust.

---

## 📂 Files to Update

- `scripts/evaluate.py`
- `src/data_preparation.py` (for training data format consistency)

---

### 1. Clarify SQL Extraction Steps in `generate_sql_batch(...)` and `prepare_prompt(...)`

Make each post-processing pattern explicit in the code, reflecting the logic below:

```python
# Common post-processing patterns:

# 1. Stop at semicolon
sql = output.split(';')[0] + ';'

# 2. Stop at newline
sql = output.split('\n')[0]

# 3. Extract using regex
sql = re.search(r'(SELECT.*?;)', output, re.IGNORECASE).group(1)

# 4. Remove markdown code blocks
sql = output.replace('```sql', '').replace('```', '')
```

Each pattern should be clearly documented and optionally toggled (via argument or flag).

---

### 2. Add EOS Token to Training Data

In `src/data_preparation.py`, ensure **assistant** responses include explicit stopping, i.e., clean SQL only — no trailing explanations.

```python
{
    "role": "assistant",
    "content": sample["answer"]  # Should end with just SQL, no extra text
}
```

This will help the model learn natural stopping behavior.

---

### 3. Improve Extraction Robustness

Handle edge cases such as:
- Multi-line SQL queries
- Extra text after SQL (e.g., explanations)
- Missing semicolons

Implement a helper function like:

```python
def extract_sql(generated_text: str) -> str:
    \"\"\"Extract SQL query from generated text.\"\"\"
    # Remove markdown code blocks if present
    text = generated_text.replace('```sql', '').replace('```', '')
    
    # Strategy 1: Stop at first semicolon
    if ';' in text:
        sql = text.split(';')[0] + ';'
        return sql.strip()
    
    # Strategy 2: Stop at common trailing phrases
    stop_phrases = ['\n\n', 'The ', 'This ', '://']
    for phrase in stop_phrases:
        if phrase in text:
            sql = text.split(phrase)[0]
            return sql.strip()
    
    # Strategy 3: Default to first line
    return text.split('\n')[0].strip()
```

---

### 4. Verify Ground Truth Format

Before evaluation, normalize both predictions and ground-truth SQL to prevent mismatches due to minor formatting differences.

```python
def normalize_sql(sql: str) -> str:
    \"\"\"Normalize SQL for fair comparison.\"\"\"
    return sql.strip().rstrip(';').lower()

correct = sum(
    1 for pred, truth in zip(predictions, ground_truths)
    if normalize_sql(pred) == normalize_sql(truth)
)
```

---

### 5. Add `max_length` Safety Constraint

Limit generation length to prevent infinite or overly long SQL generation.

```python
# Current
max_new_tokens = 256

# Consider
max_new_tokens = 128  # SQL queries are usually shorter
```

---

### 6. Minor Code Quality Notes

- **Line 99:** The `temperature` parameter in `generate_sql_batch(...)` is no longer used.  
  → Consider removing it from the function signature and call sites.

---

## ✅ Acceptance Criteria

- [ ] `generate_sql_batch(...)` and `prepare_prompt(...)` include clearly documented post-processing logic.  
- [ ] SQL extraction is robust for single-line, multi-line, and markdown-wrapped SQL.  
- [ ] Ground-truth normalization ensures consistent comparison.  
- [ ] EOS tokens included during training for clean termination.  
- [ ] `max_new_tokens` reduced to a reasonable length (≤128).  
- [ ] Obsolete parameters (e.g., temperature) removed or refactored.

---

**Assignee:** @claude  
**Labels:** `bug`, `enhancement`, `text-to-sql`, `evaluation`
